core/darwin/src/internal/TimeZoneRulesFoundation.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019-2025 JetBrains s.r.o. and contributors.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime.internal
+
+import kotlinx.cinterop.UnsafeNumber
+import kotlinx.datetime.LocalDateTime
+import kotlinx.datetime.UtcOffset
+import kotlinx.datetime.toKotlinInstant
+import kotlinx.datetime.toNSDate
+import kotlinx.datetime.toNSDateComponents
+import platform.Foundation.NSCalendar
+import platform.Foundation.NSDate
+import platform.Foundation.NSTimeZone
+import platform.Foundation.addTimeInterval
+import platform.Foundation.timeZoneWithName
+import kotlin.time.Instant
+
+internal class TimeZoneRulesFoundation(zoneId: String) : TimeZoneRules {
+    private val nsTimeZone: NSTimeZone = NSTimeZone.timeZoneWithName(zoneId)
+        ?: throw IllegalArgumentException("Unknown timezone: $zoneId")
+
+    override fun infoAtInstant(instant: Instant): UtcOffset =
+        infoAtNsDate(instant.toNSDate())
+
+    override fun infoAtDatetime(localDateTime: LocalDateTime): OffsetInfo {
+        val calendar = NSCalendar.currentCalendar().apply { timeZone = nsTimeZone }
+        val components = localDateTime.toNSDateComponents()
+        val nsDate = requireNotNull(calendar.dateFromComponents(components)) {
```

Comment:
When can this happen? In any case, it doesn't look like this could result from a user's error, so `IllegalArgumentException` doesn't seem suitable here. Instead, `IllegalStateException` (via `check`) seems better suited, and if we don't expect this to ever happen, adding more diagnostic information (`zoneId`, for one) could help us debug this if it does happen.